### ISSUE TITLE
What are the semantics of the return value and the `timeout` argument of the `pwnlib.tubes.tubes.recv*`-functions?

### ISSUE THREAD

**Issue by br0ns** [Gallopsled/pwntools#251]
The semantics have already changes several times, and I think that is mostly due to them never been explicitly written down.  So let's fix that.

Currently the semantics are as follows to the best of my knowledge.

Return value:
 - If the call was successful the received data is returned.
 - If there was a timeout ("timeout" should be clarified -- @zachriggle?) the empty string is returned.
 - If the tube is closed `EOFError` is raised.

The `timeout`-argument:
 - If `None`, the "default" value is used.  The default value can refer to a value given when creating a socket or `context.timeout` if `None` was also given here.
 - If a non-negative number is given this is used as the timeout during the function call (or the tubes lifetime if in a constructor).
 - Everything else raises an error.

I can recall at least three other semantics being applied in pwnlib.

I suggest we adapt these new semantics:

Return value:
 - If the call was successful the received data is returned.
 - If there was a timeout (again, needs clarification) `socket.timeout` is raised.
 - If the tube is closed `EOFError` is raised.

The `timeout`-argument:
 - If the string `'default'` is given, the "default" value is used, as described above.
 - If `None` is given there is no timeout, i.e. blocking forever is possible. This is also the semantics of `socket` module in the standard library.
 - If a non-negative number is given the semantics are as described above.
 - Everything else raises an error.

The biggest issue I have with the current semantics is returning the empty string when a timeout occurs.  On several occasions I've made a typo in e.g. a `recvuntil`-call and thus always got the empty string as a result; if this result is never used it can be hard to find the bug.

I also really think we should adopt the `None` means no timeout -semantics, as 1) they are used by the standard library and 2) `timeout = None` reads better.  Also note that the user will never have to actually write `'default'` so it can be anything other than `None` and numbers.

**Comment by IdolfHatler** [comment]
What's left before we can close this?

**Comment by zachriggle** [comment]
I'm good with it being closed.  I'll defer to @br0ns to finalize

**Comment by zachriggle** [comment]
Actually, the current pull request still does not throw exceptions on timeout.

**Comment by zachriggle** [comment]
So what happens when we call `recvrepeat` and it times out without actually receiving *any* data?

Still throw an exception?

**Comment by zachriggle** [comment]
Okay, so I gave a bit of a go at implementing the exception-raising behavior.  While I can *make* it work, I can't do so in a sufficiently clean manner.  Deferring this to somebody else

**Comment by zachriggle** [comment]
Can we move this to Milestone 2.3? @IdolfHatler  @br0ns

**Comment by br0ns** [comment]
OK.

**Comment by Idolf** [comment]
I have no idea what the current status of this is nor what is desirable. @zachriggle, @br0ns ?

**Comment by zachriggle** [comment]
TL;DR we still return empty strings.

```py
>>> r = remote('google.com',80)
[x] Opening connection to google.com on port 80
[x] Opening connection to google.com on port 80: Trying 2a00:1450:4009:809::200e
[+] Opening connection to google.com on port 80: Done
>>> r.recvn(1, timeout=0.1)
''
```

Given that this bug has sat stagnant for ~20 months with no progress (and I don't care enough to fix this), I'm closing this issue.
